content/blog/2019-12-09-comp-cert/index.md

```diff
@@ -0,0 +1,111 @@
+
+<img src="int.png" style="width: 40%">
+
+The live variable analysis is implemented in coq. Furthermore, the analysis is proven to generate live variable sets that are supersets of the actual live variable sets at a program point. The paper claims this is easier to prove and does not violate the correctness of the register allocation step. This is because supersets of the actual live variable sets only add more edges to the interference graph, which still maintains the correctness of the register allocation pass.
+
+The actual coloring of the interference graph is implemented in unverified OCaml code due to its complexity. The function is then used in the proof of semantic preservation using the validator approach. As a reminder, the validator approach allows CompCert to use the OCaml code only if the output is valid. Otherwise, compilation fails and no code is emitted. The correctness conditions for the coloring $\phi$ of the temporaries is:
+
+1. $\phi(r) \neq \phi(r')$ if $r$ and $r'$ interfere
+2. $\phi(r) \neq l$ if $r$ and $l$ interfere (l is a machine register or stack location. The interference graph can be pre-colored and some pseudo registers can interfere with hardware registers or stack locations)
+3. $\phi(r)$ and $r$ have the same register class (either int or float)
+
+After coloring the graph, the RTL code is transformed to LTL code by replacing temporaries according to the coloring $\phi$, which is either a hardware register or a stack location. In order to prove that the transformation preserve the semantics of the original program, the lock-step simulation approach discussed above is used. The equivalence relation on states requires that control flow is preserved, memory contents are the same, and that the registers are somehow preserved. The first two properties are intuitively correct because register allocation doesn't really affect control flow or memory state at a program point. However, the equivalence between temporaries and hardware registers is a bit more subtle. This is because the value in a hardware register might not be the same as the value in a temporary mapped to it. For example, if two temporaries that do not interfere are mapped to the same hardware register, the value in the hardware register will not be the same as one of the two temporary values at some point in time. Therefore, the paper states that a relaxation was proven where at every program point $l$, $R(r) = R'(\phi(r))$ for all $r$ live at point $l$ ($R$ is the state of the register mapping)
```

Comment:
> the coloring $\phi$, which is either a hardware register or a stack location.

Sorry to be pedantic, but $\phi(r)$ is either a hardware register or a stack location. $\phi$ itself is a mapping from temporaries to those.